File: ios/Classes/HTDrawContext.swift
//
//  HTDrawView.swift
//  Base64
//
//  Created by hublot on 2020/8/26.
//

import UIKit

class HTDrawContext {
    
    var configManager: HTKLineConfigManager
    
    weak var klineView: HTKLineView?
    

    lazy var drawItemList: [HTDrawItem] = {
        let drawItemList = [HTDrawItem]()
        return drawItemList
    }()
    
    init(_ klineView: HTKLineView, _ configManager: HTKLineConfigManager) {
        self.klineView = klineView
        self.configManager = configManager
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    var breakTouch = false
    
    func touchesGesture(_ location: CGPoint, _ translation: CGPoint, _ state: UIGestureRecognizerState) {
        guard let klineView = klineView, breakTouch == false else {
            if state == .ended {
                breakTouch = false
            }
            return
        }
        switch state {
        case .began:
            if (configManager.shouldReloadDrawItemIndex > HTDrawState.showContext.rawValue) {
                let selectedDrawItem = drawItemList[configManager.shouldReloadDrawItemIndex]
                if (selectedDrawItem.pointList.count >= selectedDrawItem.drawType.count) {
                    if (HTDrawItem.canResponseLocation(drawItemList, location, klineView) != selectedDrawItem) {
                        configManager.onDrawItemDidTouch?(nil, HTDrawState.showPencil.rawValue)
                        breakTouch = true
                        setNeedsDisplay()
                        return
                    }
                }
//            } else if (configManager.shouldReloadDrawItemIndex > HTDrawState.showPencil.rawValue) {
//                let selectedDrawItem = HTDrawItem.canResponseLocation(drawItemList, location, translation, state, klineView)
//                if let selectedDrawItem = selectedDrawItem, let selectedDrawItemIndex = drawItemList.index(of: selectedDrawItem) {
//                    configManager.onDrawItemDidTouch?(selectedDrawItem, selectedDrawItemIndex)
//                    setNeedsDisplay()
//                    return
//                } else {
//                    if HTDrawItem.canResponseTouch(drawItemList, location, translation, state, klineView) {
//                        setNeedsDisplay()
//                        return
//                    }
//                }
            }
        case .changed:
            break
        case .ended:
            break
        default:
            break
        }
        if HTDrawItem.canResponseTouch(drawItemList, location, translation, state, klineView) {
            if state == .began, let moveItem = HTDrawItem.findTouchMoveItem(drawItemList), let moveItemIndex = drawItemList.index(of: moveItem) {
                configManager.onDrawItemDidTouch?(moveItem, moveItemIndex)
            }
            setNeedsDisplay()
            return
        }
        if (configManager.drawType == .none) {
            return
        }
        
//        let moveDrawItem = HTDrawItem.findTouchMoveItem(drawItemList)
//        let canResponse = false
//        if (configManager.shouldReloadDrawItemIndex == HTDrawState.showPencil.rawValue && state == .ended && translation == CGPoint.zero) {
//            if moveDrawItem != nil {
//                configManager.shouldReloadDrawItemIndex = HTDrawState
//            }
//        }
//
//
//        // 能够处理点击, 改变拖动的点, 重新绘制
//        if let klineView = klineView, ) {
//            // 如果移动了或者点击了, 去弹起配置弹窗
//            if let moveDrawItem = moveDrawItem, let moveDrawItemIndex = drawItemList.firstIndex(of: moveDrawItem), state != .changed {
//                configManager.onDrawItemDidTouch?(moveDrawItem, moveDrawItemIndex)
//            }
//            setNeedsDisplay()
//            return
//        }
    
        
        let drawItem = drawItemList.last
        switch state {
        case .began:
            if (drawItem == nil || (drawItem?.pointList.count ?? 0) >= (drawItem?.drawType.count ?? 0)) {
                let drawItem = HTDrawItem.init(configManager.drawType, location)
                drawItem.drawColor = configManager.drawColor
                drawItem.drawLineHeight = configManager.drawLineHeight
                drawItem.drawDashWidth = configManager.drawDashWidth
                drawItem.drawDashSpace = configManager.drawDashSpace
                
                drawItemList.append(drawItem)
                configManager.onDrawItemDidTouch?(drawItem, drawItemList.count - 1)
            } else {
                drawItem?.pointList.append(location)
            }
        case .ended, .changed:
            let length = drawItem?.pointList.count ?? 0
            if length >= 1 {
                let index = length - 1
                drawItem?.pointList[index] = location
                // 最后一个点起笔
                if case .ended = state, let drawItem = drawItem {
                    configManager.onDrawPointComplete?(drawItem, drawItemList.count - 1)
                    if index == drawItem.drawType.count - 1 {
                        configManager.onDrawItemComplete?(drawItem, drawItemList.count - 1)
                        if configManager.drawShouldContinue {
                            configManager.shouldReloadDrawItemIndex = HTDrawState.showContext.rawValue
                        } else {
                            configManager.drawType = .none
                        }
                    }
                }
            }
        default:
            break
        }
        setNeedsDisplay()
    }
    
    func fixDrawItemList() {
        guard let drawItem = drawItemList.last else {
            return
        }
        if drawItem.pointList.count < drawItem.drawType.count {
            drawItemList.removeLast()
        }
        setNeedsDisplay()
    }
    
    func clearDrawItemList() {
        drawItemList = []
        setNeedsDisplay()
    }
    
    func drawLine(_ context: CGContext, _ drawItem: HTDrawItem, _ startPoint: CGPoint, _ endPoint: CGPoint) {
        context.move(to: startPoint)
        context.addLine(to: endPoint)
        context.setStrokeColor(drawItem.drawColor.cgColor)
        context.setLineWidth(drawItem.drawLineHeight)
        var dashList = [drawItem.drawDashWidth, drawItem.drawDashSpace]
        if drawItem.drawDashSpace == 0 {
            dashList = []
        }
        context.setLineDash(phase: 0, lengths: dashList)
        context.drawPath(using: .stroke)
    }
    
    func setNeedsDisplay() {
        klineView?.setNeedsDisplay()
    }

    func drawMapper(_ context: CGContext, _ drawItem: HTDrawItem, _ index: Int, _ itemIndex: Int) {
        guard let klineView = klineView else {
            return
        }
        let lineList = HTDrawItem.lineListWithIndex(drawItem, index, klineView)
        if index == 2, case .parallelLine = drawItem.drawType, let (startPoint, endPoint) = lineList.first {
            let firstPoint = drawItem.pointList[0]
            let secondPoint = drawItem.pointList[1]
            context.move(to: klineView.viewPointFromValuePoint(firstPoint))
            context.addLine(to: klineView.viewPointFromValuePoint(secondPoint))
            context.addLine(to: klineView.viewPointFromValuePoint(startPoint))
            context.addLine(to: klineView.viewPointFromValuePoint(endPoint))
            context.closePath()
            context.setFillColor(drawItem.drawColor.withAlphaComponent(0.5).cgColor)
            context.drawPath(using: .fill)
            let dashStartPoint = HTDrawItem.centerPoint(p1: firstPoint, p2: endPoint)
            let dashEndPoint = HTDrawItem.centerPoint(p1: secondPoint, p2: startPoint)
            context.move(to: klineView.viewPointFromValuePoint(dashStartPoint))
            context.addLine(to: klineView.viewPointFromValuePoint(dashEndPoint))
            context.setLineDash(phase: 0, lengths: [4, 4])
            context.setStrokeColor(drawItem.drawColor.withAlphaComponent(0.5).cgColor)
            context.setLineWidth(1)
            context.drawPath(using: .stroke)
        }
        for (startPoint, endPoint) in lineList {
            drawLine(context, drawItem, klineView.viewPointFromValuePoint(startPoint), klineView.viewPointFromValuePoint(endPoint))
        }
        let point = drawItem.pointList[index]
        
        if (itemIndex != configManager.shouldReloadDrawItemIndex) {
            return
        }
        
        context.addArc(center: klineView.viewPointFromValuePoint(point), radius: 10, startAngle: 0, endAngle: CGFloat(Double.pi * 2.0), clockwise: true)
        context.setFillColor(drawItem.drawColor.withAlphaComponent(0.5).cgColor)
        context.drawPath(using: .fill)
        context.addArc(center: klineView.viewPointFromValuePoint(point), radius: 4, startAngle: 0, endAngle: CGFloat(Double.pi * 2.0), clockwise: true)
        context.setFillColor(drawItem.drawColor.cgColor)
        context.drawPath(using: .fill)
    }
    
    func draw(_ contenOffset: CGFloat) {
        guard let context = UIGraphicsGetCurrentContext() else {
            return
        }
        for (itemIndex, drawItem) in drawItemList.enumerated() {
            for (index, _) in drawItem.pointList.enumerated() {
                drawMapper(context, drawItem, index, itemIndex)
            }
        }
    }

}

File: android/src/main/java/com/github/fujianlian/klinechart/container/HTDrawContext.java
package com.github.fujianlian.klinechart.container;

import android.content.Context;
import android.graphics.*;
import android.view.MotionEvent;
import android.view.View;
import com.github.fujianlian.klinechart.BaseKLineChartView;
import com.github.fujianlian.klinechart.HTKLineConfigManager;
import com.github.fujianlian.klinechart.KLineChartView;

import java.util.ArrayList;
import java.util.List;

public class HTDrawContext {

    public List<HTDrawItem> drawItemList = new ArrayList<HTDrawItem>();

    private Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private BaseKLineChartView klineView;

    private HTKLineConfigManager configManager;

    private Boolean breakTouch = false;

    public HTDrawContext(BaseKLineChartView klineView, HTKLineConfigManager configManager) {
        this.klineView = klineView;
        this.configManager = configManager;
    }

    public void touchesGesture(HTPoint location, HTPoint translation, int state) {
        // 能够处理点击, 改变拖动的点, 重新绘制
        if (breakTouch == true) {
            if (state == MotionEvent.ACTION_UP) {
                breakTouch = false;
            }
            return;
        }
        switch (state) {
            case MotionEvent.ACTION_DOWN: {
                if (configManager.shouldReloadDrawItemIndex > HTDrawState.showContext) {
                    HTDrawItem selectedDrawItem = drawItemList.get(configManager.shouldReloadDrawItemIndex);
                    if (selectedDrawItem.pointList.size() >= selectedDrawItem.drawType.count()) {
                        if (HTDrawItem.canResponseLocation(drawItemList, location, klineView) != selectedDrawItem) {
                            configManager.onDrawItemDidTouch.invoke(null, HTDrawState.showPencil);
                            breakTouch = true;
                            invalidate();
                            return;
                        }
                    }
                }
                break;
            }
        }
        if (HTDrawItem.canResponseTouch(drawItemList, location, translation, state, klineView)) {
            if (state == MotionEvent.ACTION_DOWN) {
                HTDrawItem moveItem = HTDrawItem.findTouchMoveItem(drawItemList);
                if (moveItem != null) {
                    int moveItemIndex = drawItemList.indexOf(moveItem);
                    configManager.onDrawItemDidTouch.invoke(moveItem, moveItemIndex);
                }
            }
            invalidate();
            return;
        }
        if (configManager.drawType == HTDrawType.none) {
            return;
        }


        int size = drawItemList.size();
        HTDrawItem drawItem = size > 0 ? drawItemList.get(size - 1) : null;
        switch (state) {
            case MotionEvent.ACTION_DOWN:
            if (drawItem == null || (drawItem.pointList.size() >= drawItem.drawType.count())) {
                drawItem = new HTDrawItem(configManager.drawType, location);
                drawItem.drawColor = configManager.drawColor;
                drawItem.drawLineHeight = configManager.drawLineHeight;
                drawItem.drawDashWidth = configManager.drawDashWidth;
                drawItem.drawDashSpace = configManager.drawDashSpace;
                drawItemList.add(drawItem);
                configManager.onDrawItemDidTouch.invoke(drawItem, drawItemList.size() - 1);
            } else {
                drawItem.pointList.add(location);
            }
            case MotionEvent.ACTION_MOVE: {

            }
            case MotionEvent.ACTION_UP:
                int length = drawItem.pointList.size();
                if (length >= 1) {
                int index = length - 1;
                drawItem.pointList.set(index, location);
                if (state == MotionEvent.ACTION_UP) {
                    configManager.onDrawPointComplete.invoke(drawItem, drawItemList.size() - 1);
                    if (index == drawItem.drawType.count() - 1) {
                        configManager.onDrawItemComplete.invoke(drawItem, drawItemList.size() - 1);
                        if (configManager.drawShouldContinue) {
                            configManager.shouldReloadDrawItemIndex = HTDrawState.showContext;
                        } else {
                            configManager.drawType = HTDrawType.none;
                        }
                    }
                }
            }
            default:
                break;
        }
        invalidate();
    }

    public void invalidate() {
        klineView.invalidate();
    }

    public void fixDrawItemList() {
        int size = drawItemList.size();
        if (size <= 0) {
            return;
        }
        HTDrawItem drawItem = drawItemList.get(size - 1);
        if (drawItem.pointList.size() < drawItem.drawType.count()) {
            drawItemList.remove(drawItem);
        }
        invalidate();
    }

    public void clearDrawItemList() {
        drawItemList = new ArrayList<>();
        invalidate();
    }

    private int colorWithAlphaComponent(int color, double alpha) {
        int reloadColor = (color & 0x00FFFFFF) | ((int)(alpha * 255) << 24);
        return reloadColor;
    }

    private void drawLine(Canvas canvas, HTDrawItem drawItem, HTPoint startPoint, HTPoint endPoint) {
        paint.setColor(drawItem.drawColor);
        paint.setPathEffect(new DashPathEffect(new float[] { drawItem.drawDashWidth, drawItem.drawDashSpace }, 0));
        paint.setStrokeWidth(drawItem.drawLineHeight);
        Path path = new Path();
        path.moveTo(startPoint.x, startPoint.y);
        path.lineTo(endPoint.x, endPoint.y);
        paint.setStyle(Paint.Style.STROKE);
        canvas.drawPath(path, paint);
    }

    public void drawMapper(Canvas canvas, HTDrawItem drawItem, int index, int itemIndex) {
        HTPoint point = drawItem.pointList.get(index);
        List<List<HTPoint>> lineList = HTDrawItem.lineListWithIndex(drawItem, index, klineView);
        if (index == 2 && drawItem.drawType == HTDrawType.parallelLine) {
            List<HTPoint> firstLine = lineList.get(0);
            HTPoint startPoint = firstLine.get(0);
            HTPoint endPoint = firstLine.get(1);
            HTPoint firstPoint = drawItem.pointList.get(0);
            HTPoint secondPoint = drawItem.pointList.get(1);
            Path path = new Path();

            HTPoint firstViewPoint = klineView.viewPointFromValuePoint(firstPoint);
            HTPoint secondViewPoint = klineView.viewPointFromValuePoint(secondPoint);
            HTPoint startViewPoint = klineView.viewPointFromValuePoint(startPoint);
            HTPoint endViewPoint = klineView.viewPointFromValuePoint(endPoint);

            path.moveTo(firstViewPoint.x, firstViewPoint.y);
            path.lineTo(secondViewPoint.x, secondViewPoint.y);
            path.lineTo(startViewPoint.x, startViewPoint.y);
            path.lineTo(endViewPoint.x, endViewPoint.y);
            path.close();
            paint.setColor(colorWithAlphaComponent(drawItem.drawColor, 0.5));
            paint.setStyle(Paint.Style.FILL);
            canvas.drawPath(path, paint);

            HTPoint dashStartPoint = HTDrawItem.centerPoint(firstPoint, endPoint);
            HTPoint dashEndPoint = HTDrawItem.centerPoint(secondPoint, startPoint);
            path = new Path();

            HTPoint dashStartViewPoint = klineView.viewPointFromValuePoint(dashStartPoint);
            HTPoint dashEndViewPoint = klineView.viewPointFromValuePoint(dashEndPoint);

            path.moveTo(dashStartViewPoint.x, dashStartViewPoint.y);
            path.lineTo(dashEndViewPoint.x, dashEndViewPoint.y);
            paint.setColor(colorWithAlphaComponent(drawItem.drawColor, 0.5));
            paint.setPathEffect(new DashPathEffect(new float[] { 4, 4 }, 0));
            paint.setStyle(Paint.Style.STROKE);
            paint.setStrokeWidth(2);
            canvas.drawPath(path, paint);
        }
        for (List<HTPoint> pointList: lineList) {
            HTPoint startPoint = pointList.get(0);
            HTPoint endPoint = pointList.get(1);
            drawLine(canvas, drawItem, klineView.viewPointFromValuePoint(startPoint), klineView.viewPointFromValuePoint(endPoint));
        }

        if (itemIndex != configManager.shouldReloadDrawItemIndex) {
            return;
        }



        HTPoint viewPoint = klineView.viewPointFromValuePoint(point);

        Path path = new Path();
        paint.setStyle(Paint.Style.FILL);
        path.addCircle(viewPoint.x, viewPoint.y, 20, Path.Direction.CW);
        paint.setColor(colorWithAlphaComponent(drawItem.drawColor, 0.5));
        canvas.drawPath(path, paint);

        path = new Path();
        paint.setStyle(Paint.Style.FILL);
        path.addCircle(viewPoint.x, viewPoint.y, 8, Path.Direction.CW);
        paint.setColor(drawItem.drawColor);
        canvas.drawPath(path, paint);
    }


    public void onDraw(Canvas canvas) {
        for (int itemIndex = 0; itemIndex < drawItemList.size(); itemIndex ++) {
            HTDrawItem drawItem = drawItemList.get(itemIndex);
            for (int index = 0; index < drawItem.pointList.size(); index ++) {
                drawMapper(canvas, drawItem, index, itemIndex);
            }
        }
    }
}
